### PR TITLE
Ensure all `{modelIdentity}/{action}` actions are handled as 'model-related'

### PR DESCRIPTION
**Pull request by danielsharvey**
### Overview

Ensure all `{modelIdentity}/{action}` actions are handled as 'model-related':

1. All `{modelIdentity}/{action}` actions treated as related to the model (including `/allActions`) and Swagger content integrated / merged appropriately.
2. Actions2 actions recognised as model-related, and therefore recognised as blueprint action overrides (where applicable)
3. Improve consistency of handling/merging of actions (and Swagger documentation) relating to model controllers.

### Details

- refactor: Improve `SwaggerRouteInfo` interface

    Improve `SwaggerRouteInfo` interface to support above:
    1. Add separate fields for action vs blueprint action.
    2. Add separate flag for shortcut blueprint routes.

- refactor: Refactor route parsing using `{modelIdentity}/{action}`
    
    Refactor route parsing to rely predominantly on `{modelIdentity}/{action}` to ensure all model-related actions (including overrides including actions2) are captured. This is consistent with Sails `parseBlueprintOptions()`.

- refactor: Refactor transformations phase for `SwaggerRouteInfo` changes

- refactor: Refactor Swagger generation for model-related changes

    Refactor generation to improve route/model/controller Swagger merge.
    
    Refactor for `SwaggerRouteInfo` changes. 

- test: Refactor tests for Swagger generation (model-related changes)

    Swagger generation  changes related to new test cases and model-related merge changes (UserController now equivalent to User model).    

    Swagger generation merge changes resulted in ordering changes in generated output.
    
    Add new Actions2 action overriding blueprint action.
    
    Adjust `UserController` to add new Swagger merge.
    
    Includes full re-generation of parsed routes based on `SwaggerRouteInfo` changes.
    
    Minor adjustment to Sails AttributeDefinition as `type` optional.

### CODE REVIEW

**Comment by danielsharvey** [comment]
Closes #81.

**Comment by danielsharvey** [comment]
@theoomoregbee A review would be good for this change. The way `{modelIdentity}/{action}` actions were parsed meant that actions2 actions which override default blueprint action were not treated as such during Swagger generation. I've reviewed and re-worked this flow. This has resulted in a few other side-effects (fixes, but somewhat subtle) in the way controller vs model swagger is merged. I've checked the resulting changes to the test Swagger generated carefully (you can follow through the individual commits on this) but would be good to have a second opinion.

**Comment by theoomoregbee** [comment]
cool! will give this a look before end of today. Thanks @danielsharvey

File: lib/swagger-doc.ts [hunk 1]
```diff
@@ -64,7 +64,7 @@ export default async (sails: Sails.Sails, sailsRoutes: Array<Sails.Route>, conte
    * @see https://github.com/balderdashy/sails/blob/master/lib/hooks/blueprints/index.js#L401
    */
   if(hookConfig.excludeDeprecatedPutBlueprintRoutes) {
-    routes = routes.filter(route => !(route.actionType === 'blueprint' && route.action === 'update' && route.verb === 'put'));
+    routes = routes.filter(route => !(route.model && route.blueprintAction === 'update' && route.verb === 'put'));
```
[the remark this answers is not in the document]
Comment (danielsharvey) [reply lib/swagger-doc.ts:67]: You're right, this check is superfluous.

File: lib/parsers.ts [hunk 2]
```diff
@@ -275,80 +280,66 @@ export const parseBoundRoutes = (boundRoutes: Array<Sails.Route>,
...
-    }
+      // fall-through --> non-model based action
 
-  })
-  .filter(route => !!route) as SwaggerRouteInfo[];
+      return {
+        middlewareType,
+        verb: verb as HTTPMethodVerb,
+        ...parsedPath, // path & variables
+        action: routeOptions.action || mwtAction || '_unknown',
```
[the remark this answers is not in the document]
Comment (danielsharvey) [reply lib/parsers.ts:336]: This worth a check - I saw it as a fail-safe that would at least indicate where to look if action was undefined, but it is probably a very, very minor edge case.

**Pull request merged**